enterprise/backend/src/metabase_enterprise/task/cache.clj

```diff
@@ -43,23 +45,37 @@
   [refresh-task-fn]
   (.submit ^ExecutorService @pool ^Callable refresh-task-fn))
 
+(defn discarding-rff
+  "Returns a reducing function that discards result rows"
+  [_metadata]
+  (fn discarding-rf
+    ([] {:rows []})
+    ([result] result)
+    ([result _row] result)))
+
 (defn- refresh-task
   "Returns a function that serially reruns queries based on `refresh-defs`, and discards the results. Each refresh
   definition contains a card-id, an optional dashboard-id, and a list of queries to rerun."
   [refresh-defs]
   (fn []
-    (doseq [{:keys [card-id dashboard-id queries]} refresh-defs
-            query queries]
-      (try
-        (qp/process-query
-         (qp/userland-query
-          (assoc-in query [:middleware :ignore-cached-results?] true)
-          {:executed-by  nil
-           :context      :cache-refresh
-           :card-id      card-id
-           :dashboard-id dashboard-id}))
-        (catch Exception e
-          (log/debugf "Error refreshing cache for card %s: %s" card-id (ex-message e)))))))
+    (doseq [{:keys [card-id dashboard-id queries]} refresh-defs]
+      ;; Annotate the query with its cache strategy in the format expected by the QP
+      (let [cache-strategy (strategies/cache-strategy (t2/select-one :model/Card :id card-id)
```

Comment:
I wonder if this DB call can be eliminated. It could certainly be replaced by single query with the set of `card-id`s, but the caller might actually have both the card and the query.

Reply:
The caller doesn't actually have the card already, but good call about using a single query. Changing that.